File: benchmarks/executable/restaurant_uint_bitwise.c
// C23 reference for the Restaurant UInt binary-bitwise workload.

#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

static volatile uint64_t runtime_high = UINT64_C(9223372036854775808);
static volatile uint64_t runtime_lower = UINT64_C(9223372036854775807);

int main(void) {
#ifdef _WIN32
    if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif
    const uint64_t left = runtime_high;
    const uint64_t right = runtime_lower;
    const uint64_t combined = left | (right ^ (left & right));
    return printf("UInt bits %" PRIu64 "\n", combined) < 0 ? 1 : 0;
}

File: benchmarks/executable/restaurant_uint_bitwise.rs
// Rust 2024 reference for the Restaurant UInt binary-bitwise workload.

use std::hint::black_box;
use std::io::{self, Write};

fn main() {
    let left = black_box(9_223_372_036_854_775_808_u64);
    let right = black_box(9_223_372_036_854_775_807_u64);
    let combined = left | (right ^ (left & right));
    let mut stdout = io::stdout().lock();
    if writeln!(stdout, "UInt bits {combined}").is_err() {
        std::process::exit(1);
    }
}
